contrib/babelfishpg_tsql/src/pltsql_bulkcopy.c

```diff
@@ -0,0 +1,938 @@
+/*-------------------------------------------------------------------------
+ *
+ * pltsql_bulkcopy.c		- Bulk Copy for PL/tsql
+ *
+ * IDENTIFICATION
+ *	  contrib/babelfishpg_tsql/src/pltsql_bulkcopy.c
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include <ctype.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include "access/heapam.h"
+#include "access/tableam.h"
+#include "access/xact.h"
+#include "catalog/dependency.h"
+#include "catalog/namespace.h"
+#include "commands/sequence.h"
+#include "executor/executor.h"
+#include "executor/nodeModifyTable.h"
+#include "executor/tuptable.h"
+#include "optimizer/optimizer.h"
+#include "miscadmin.h"
+#include "parser/parse_relation.h"
+#include "pltsql_bulkcopy.h"
+#include "rewrite/rewriteHandler.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/rel.h"
+#include "utils/rls.h"
+
+/*
+ * No more than this many tuples per CopyMultiInsertBuffer
+ *
+ * Caution: Don't make this too big, as we could end up with this many
+ * CopyMultiInsertBuffer items stored in CopyMultiInsertInfo's
+ * multiInsertBuffers list.
+ */
+#define MAX_BUFFERED_TUPLES		1000
+
+/* Trim the list of buffers back down to this number after flushing */
+#define MAX_PARTITION_BUFFERS	32
+
+/* Stores multi-insert data related to a single relation. */
+typedef struct CopyMultiInsertBuffer
```

Comment:
How much is this code different from PG copy? Did we consider reusing PG functions here?

Reply:
We have re-used the static functions which start with CopyMultiInsert (with few changes)
Things to note:
1. PG COPY functionality is way different since it reads from a file.
2. These functions are static to PG COPY and also provide additional help in maintaining the state for reading from files. (In our design discussion we brought this up and it was agreed that for now the best would be to add comments for the functions re-used)